src/nestegg.c

```diff
@@ -1553,6 +1561,36 @@ ne_read_block(nestegg * ctx, uint64_t block_id, uint64_t block_size, nestegg_pac
         }
         f->frame_encryption->length = IV_SIZE;
         encryption_size = SIGNAL_BYTE_SIZE + IV_SIZE;
+
+        if ((signal_byte & PARTITIONED_BIT_MASK) == PACKET_PARTITIONED) {
+          r = ne_io_read(ctx->io, &f->frame_encryption->num_partitions, NUM_PACKETS_SIZE);
+          if (r != 1) {
+            free(f->frame_encryption->iv);
+            free(f->frame_encryption);
+            free(f);
+            nestegg_free_packet(pkt);
+            return r;
+          }
+          
+          encryption_size += NUM_PACKETS_SIZE + f->frame_encryption->num_partitions * PACKET_OFFSET_SIZE;
+          f->frame_encryption->partition_offsets = ne_alloc(f->frame_encryption->num_partitions * PACKET_OFFSET_SIZE);
+          r = 0;
+          for (i = 0; i < f->frame_encryption->num_partitions; i++) {
+            uint64_t value = 0;
+            r += ne_read_uint(ctx->io, &value, PACKET_OFFSET_SIZE);
+            f->frame_encryption->partition_offsets[i] = (uint32_t)value;
+          }
+
+          /* If any of the partition offsets did not return 1, then fail. */
+          if (r != f->frame_encryption->num_partitions) {
+            free(f->frame_encryption->iv);
+            free(f->frame_encryption);
+            free(f->frame_encryption->partition_offsets);
```

Comment:
Use after free here.

To reduce the future risk and some code duplication, I think we should replace this (and all other) chunk(s) of frees of `f` and `frame_encryption` (and sub-fields)  with a helper function.  Add a `alloc_frame` that replaces the allocation and initial NULL/zeroing of `struct frame` and a `free_frame` that does something like:

    if (f->frame_encryption) {
      free(f->frame_encryption->iv);
      free(f->frame_encryption->partition_offsets);
    }
    free(f->frame_encryption);
    free(f->data);
    free(f);